src/ripple/app/consensus/RCLValidations.cpp

```diff
@@ -0,0 +1,254 @@
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <BeastConfig.h>
+#include <ripple/app/consensus/RCLValidations.h>
+#include <ripple/app/ledger/LedgerMaster.h>
+#include <ripple/app/main/Application.h>
+#include <ripple/app/misc/NetworkOPs.h>
+#include <ripple/app/misc/ValidatorList.h>
+#include <ripple/basics/Log.h>
+#include <ripple/basics/StringUtilities.h>
+#include <ripple/basics/chrono.h>
+#include <ripple/consensus/LedgerTiming.h>
+#include <ripple/core/DatabaseCon.h>
+#include <ripple/core/JobQueue.h>
+#include <ripple/core/TimeKeeper.h>
+#include <memory>
+#include <mutex>
+#include <thread>
+
+namespace ripple {
+
+RCLValidationsPolicy::RCLValidationsPolicy(Application& app) : app_(app)
+{
+    staleValidations_.reserve(512);
+}
+
+NetClock::time_point
+RCLValidationsPolicy::now() const
+{
+    return app_.timeKeeper().closeTime();
+}
+
+void
+RCLValidationsPolicy::onStale(RCLValidation&& v)
+{
+    // Store the newly stale validation; do not do significant work in this
+    // function since this is a callback from Validations, which may be
+    // doing other work.
+
+    ScopedLockType sl(staleLock_);
+    staleValidations_.emplace_back(std::move(v));
+    if (staleWriting_)
+        return;
+
+    staleWriting_ = true;
+    app_.getJobQueue().addJob(
+        jtWRITE, "Validations::doStaleWrite", [this](Job&) {
+            auto event =
+                app_.getJobQueue().getLoadEventAP(jtDISK, "ValidationWrite");
+            ScopedLockType sl(staleLock_);
+            doStaleWrite(sl);
+        });
+}
+
+void
+RCLValidationsPolicy::flush(hash_map<PublicKey, RCLValidation>&& remaining)
+{
+    bool anyNew = false;
+    {
+        ScopedLockType sl(staleLock_);
+
+        for (auto const& keyVal : remaining)
+        {
+            staleValidations_.emplace_back(std::move(keyVal.second));
+            anyNew = true;
+        }
+
+        // If we have new validations to write and there isn't a write in
+        // progress already, then write to the database synchronously.
+        if (anyNew && !staleWriting_)
+        {
+            staleWriting_ = true;
+            doStaleWrite(sl);
+        }
+
+        // Handle the case where flush() is called while a queuedWrite
+        // is already in progress.
```

Comment:
This was a port of the changes in #2083.  If `flush` is called while an existing async job was queued or is now running, this would block until it completes.  It would be nice to drop this if we don't care about whether that job finishes before returning.

@scottschurr: What is your opinion?  Is it important for the `flush` call to block until the potentially asynchronous call is complete?